### PR TITLE
OpenCAS Linux kernel 5.15 adaptation

### PR DESCRIPTION
**Pull request by kmajzero**
This patch adapts OpenCAS Linux to compile and work with kernel v5.15

Signed-off-by: Krzysztof Majzerowicz-Jaszcz <krzysztof.majzerowicz-jaszcz@intel.com>

### CODE REVIEW

File: modules/cas_disk/exp_obj.c [hunk 1]
```diff
@@ -530,7 +529,12 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
 	exp_obj->owner = owner;
 	exp_obj->ops = ops;
 
-	gd = alloc_disk(1);
+	result = _casdsk_init_tag_set(dsk, &dsk->tag_set);
+	if (result) {
+		goto error_init_tag_set;
+	}
+
+	gd = blk_mq_alloc_disk(&dsk->tag_set, NULL);
```
Comment (robertbaldyga) [inline modules/cas_disk/exp_obj.c:537]: This needs additional build config, otherwise it will break building on older kernels.

File: modules/cas_disk/exp_obj.c [hunk 2]
```diff
@@ -701,7 +692,7 @@ int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
 	if (!exp_obj->locked_bd)
 		return -ENAVAIL;
 
-	mutex_lock(&exp_obj->locked_bd->bd_mutex);
+	mutex_lock(&exp_obj->locked_bd->bd_disk->open_mutex);
```
Comment (robertbaldyga) [inline modules/cas_disk/exp_obj.c:695]: Same here.

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

**Comment by mmichal10** [comment]
How about including hashes of commits (from the kernel code) which introduced the changes? Having such an information in the commit message would made review easier for people who are not up to date with the recent changes in the kernel.

File: modules/cas_disk/exp_obj.c [hunk 3]
```diff
@@ -575,8 +568,6 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
 
 error_set_geometry:
 	blk_cleanup_queue(exp_obj->queue);
-error_init_queue:
-	blk_mq_free_tag_set(&dsk->tag_set);
 error_init_tag_set:
 	_casdsk_exp_obj_clear_dev_t(dsk);
 error_dev_t:
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:573]: As you reordered the operations, you should also adjust error handling to the changes. Additionally, I'm not sure if we can simply get rid of `blk_mq_free_tag_set()` as `_casdsk_init_tag_set()` is still present in the code and as far as I see it wasn't affected by changes in the kernel.

File: modules/cas_disk/exp_obj.c [hunk 4]
```diff
@@ -530,36 +530,29 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
 	exp_obj->owner = owner;
 	exp_obj->ops = ops;
 
-	gd = alloc_disk(1);
-	if (!gd) {
-		result = -ENOMEM;
+	result = _casdsk_init_tag_set(dsk, &dsk->tag_set);
+	if (result) {
+		goto error_init_tag_set;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:535]: Since the module ref count has been incremented a few lines above, you need to call `module_put()` in rollback

File: modules/cas_disk/exp_obj.c [hunk 5]
```diff
@@ -575,15 +568,14 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
-error_init_tag_set:
-	_casdsk_exp_obj_clear_dev_t(dsk);
 error_dev_t:
 	put_disk(gd);
 error_alloc_disk:
 	module_put(owner);
 	dsk->exp_obj->owner = NULL;
+	blk_mq_free_tag_set(&dsk->tag_set);
+error_init_tag_set:
+	_casdsk_exp_obj_clear_dev_t(dsk);
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:578]: `_casdsk_exp_obj_set_dev_t()` is the last but one function that might return an error therefore `_casdsk_exp_obj_clear_dev_t()` should be called as a rollback for `error_set_geometry` (which is the last error)

File: modules/cas_disk/exp_obj.c [hunk 6]
```diff
@@ -530,36 +530,29 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
+	result = cas_alloc_mq_disk(&gd, &queue, &dsk->tag_set);
+	if (result) {
 		goto error_alloc_disk;
 	}
+
 	exp_obj->gd = gd;
 
 	result = _casdsk_exp_obj_set_dev_t(dsk, gd);
 	if (result)
 		goto error_dev_t;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:547]: Since `cas_alloc_mq_disk()` succeeded a few lines above, you need to call `blk_cleanup_queue()` in rollback

**Review by mmichal10 — COMMENTED** [review]
I strongly suggest to avoid reusing the existing rollback (especially the existing labels). I think the easier solution would be to remove whole rollback procedure and create it from scratch.

File: modules/cas_disk/exp_obj.c [hunk 7]
```diff
@@ -339,8 +339,32 @@ static void _casdsk_exp_obj_clear_dev_t(struct casdsk_disk *dsk)
 	}
 }
 
+static int _casdsk_exp_obj_open(struct block_device *bdev, fmode_t mode)
+{
+	struct casdsk_disk *dsk = bdev->bd_disk->private_data;
+
+	mutex_lock(&dsk->lock);
```
Comment (arutk) [inline modules/cas_disk/exp_obj.c:346]: there is a wrapper for that: casdsk_disk_lock()

Comment (arutk) [reply modules/cas_disk/exp_obj.c:346]: I'm not sure how this interferes with other usages of casdsk_disk_lock(). For example casdsk_disk_set_pt(). I think almost all of the functions that currently call  casdsk_disk_lock() are not used at all. I suggest to remove them and check whether dsk->lock remains used anywhere else at all.

**Review by arutk — CHANGES_REQUESTED** [review]

File: modules/cas_disk/exp_obj.c [hunk 8]
```diff
@@ -396,6 +419,7 @@ EXPORT_SYMBOL(casdsk_exp_obj_free);
 
 static void __casdsk_exp_obj_release(struct casdsk_exp_obj *exp_obj)
 {
+	//kfree(exp_obj->gd->private_data);
```
Comment (arutk) [inline modules/cas_disk/exp_obj.c:422]: why?

File: modules/cas_disk/disk.c [hunk 9]
```diff
@@ -249,9 +249,9 @@ void __exit casdsk_disk_shutdown_all(void)
 		atomic_set(&dsk->mode, CASDSK_MODE_SHUTDOWN);
 
 		if (dsk->exp_obj) {
-			casdsk_exp_obj_lock(dsk);
+			//casdsk_exp_obj_lock(dsk);
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply modules/cas_disk/disk.c:252]: this is due to my error while applying changes. New version fixes this and others

[the remark this answers is not in the document]
Comment (arutk) [reply modules/cas_disk/disk.c:252]: you still have this commented

**Review by arutk — CHANGES_REQUESTED** [review]

File: modules/cas_disk/exp_obj.c [hunk 10]
```diff
@@ -505,61 +505,54 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
 		__casdsk_exp_obj_release(exp_obj);
-		goto error_kobject;
+		goto error_init_kobject;
 	}
 
 	if (!try_module_get(owner)) {
 		CASDSK_DEBUG_DISK_ERROR(dsk, "Cannot get reference to module");
 		result = -ENAVAIL;
-		goto error_module;
+		goto error_module_get;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:558]: In case of failing to increment module reference count or any further errors `__casdsk_exp_obj_release()` won't be called

File: modules/cas_disk/exp_obj.c [hunk 11]
```diff
@@ -683,6 +713,7 @@ EXPORT_SYMBOL(casdsk_exp_obj_activated);
 int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
 {
 	struct casdsk_exp_obj *exp_obj;
+	int result = 0;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:716]: Unused variable

File: modules/cas_disk/exp_obj.c [hunk 12]
```diff
@@ -691,36 +722,29 @@ int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
...
 
+	if (dsk->openers > 0) {
+		mutex_unlock(&dsk->openers_lock);
 		casdsk_exp_obj_unlock(dsk);
 		return -EBUSY;
 	}
 
-	return 0;
+	dsk->claimed = true;
+	mutex_unlock(&dsk->openers_lock);
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:734]: How about adding wrappers for claiming and disclaiming the exported object?

File: modules/cas_disk/disk.h [hunk 13]
```diff
@@ -31,6 +31,10 @@ struct casdsk_disk {
 
 	struct mutex lock;
 
+	struct mutex openers_lock;
```
Comment (mmichal10) [inline modules/cas_disk/disk.h:34]: You need to initialize this mutex before using it

**Review by arutk — CHANGES_REQUESTED** [review]

File: configure.d/1_alloc_disk.conf [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+#
+
+. $(dirname $3)/conf_framework
+
+
+check() {
+	cur_name=$(basename $2)
+	config_file_path=$1
+	if compile_module $cur_name "blk_mq_alloc_disk(NULL, NULL);" "linux/blk-mq.h"
+    then
```
Comment (arutk) [inline configure.d/1_alloc_disk.conf:14]: please fix alignment

File: configure.d/1_reread_partitions.conf [hunk 15]
```diff
@@ -23,7 +23,7 @@ apply() {
         add_function "
         static inline void cas_reread_partitions(struct block_device *bdev)
         {
-            bdev_disk_changed(bdev, false);
+            bdev_disk_changed(bdev->bd_disk, false);
```
Comment (arutk) [inline configure.d/1_reread_partitions.conf:26]: is this going to work on older kernels?

File: modules/cas_disk/exp_obj.c [hunk 16]
```diff
@@ -697,36 +722,29 @@ int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
...
-
-	if (exp_obj->locked_bd->bd_openers) {
-		printk(CASDSK_KERN_DEBUG "Device %s in use (openers=%d). Refuse to stop\n",
-		       exp_obj->locked_bd->bd_disk->disk_name,
-		       exp_obj->locked_bd->bd_openers);
+	mutex_lock(&dsk->openers_lock);
 
+	if (dsk->openers > 0) {
+		mutex_unlock(&dsk->openers_lock);
 		casdsk_exp_obj_unlock(dsk);
```
Comment (arutk) [inline modules/cas_disk/exp_obj.c:729]: unnecessary and potentially racy

File: modules/cas_disk/exp_obj.c [hunk 17]
```diff
@@ -505,61 +535,54 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
+		goto error_exp_obj_alloc;
 
 	exp_obj = dsk->exp_obj;
 
 	exp_obj->dev_name = kstrdup(dev_name, GFP_KERNEL);
 	if (!exp_obj->dev_name) {
 		__casdsk_exp_obj_release(exp_obj);
 		result = -ENOMEM;
-		goto error_strdup;
+		goto error_kstrdup;
```
Comment (arutk) [inline modules/cas_disk/exp_obj.c:546]: jumping to error_kstrdup will call casdsk_exp_obj_free and put the kobject ("kobject_put(&exp_obj->kobj)") while the kobject is initialized later in _casdsk_exp_obj_init_kobject().

**Review by arutk — CHANGES_REQUESTED** [review]
Error handling in exported object initialization is very complex and hard to follow. I am wondering about splitting this function into a chain of smaller functions, calling each other sequentially, each with its own smaller error handling. This might be artificial, but would improve readability. Please think about this.

I mean a structure like this, with the main init function split into f1(), f2(), ... etc:

`f1() {
	err = do_1_1()
	if (err) 
		goto cleanup1_1
	err = do_1_2()
	if (err)
		goto cleanup1_2

	err = f2()
	if (!err)
		return 0;

	cleanup1_2:
	...
	cleanup1_1:
	...
	
	return err
}`

File: modules/cas_disk/disk.h (hunk 13 above)
Comment (kmajzero) [reply modules/cas_disk/disk.h:34]: it is initalized in disk.c:129, right after dsk->lock initialization.

File: modules/cas_disk/disk.c [hunk 18]
```diff
@@ -126,6 +126,8 @@ struct casdsk_disk *casdsk_disk_open(const char *path, void *private)
 	}
 	mutex_init(&dsk->lock);
 
+	mutex_init(&dsk->openers_lock);
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply modules/cas_disk/disk.c:129]: no, mutex deinit/destroy is not needed in the kernel, as there is no resources to free

File: modules/cas_disk/exp_obj.c [hunk 19]
```diff
@@ -339,8 +339,39 @@ static void _casdsk_exp_obj_clear_dev_t(struct casdsk_disk *dsk)
...
+	mutex_unlock(&dsk->openers_lock);
+	return result;
+}
+
+static void _casdsk_exp_obj_close(struct gendisk *gd, fmode_t mode)
+{
+	struct casdsk_disk *dsk = gd->private_data;
+
+	mutex_lock(&dsk->openers_lock);
+	dsk->openers--;
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c:366]: good point, done!

File: configure.d/1_reread_partitions.conf (hunk 15 above)
Comment (kmajzero) [reply configure.d/1_reread_partitions.conf:26]: It does, configure script would use "ioctl_by_bdev" implementation on 5.4 and earlier kernels.

File: modules/cas_disk/exp_obj.c (hunk 12 above)
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c]: Basically, cas_exp_obj_lock/unlock are wrappers for claiming and disclaiming the exported object.

File: modules/cas_disk/exp_obj.c [hunk 20]
```diff
@@ -339,8 +339,41 @@ static void _casdsk_exp_obj_clear_dev_t(struct casdsk_disk *dsk)
...
+		result = -ENAVAIL;
+		goto error_claimed;
+	}
+
+	dsk->openers++;
+
+error_claimed:
+	mutex_unlock(&dsk->openers_lock);
+	return result;
+}
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c:359]: true, good catch, thanks

File: modules/cas_disk/exp_obj.c [hunk 21]
```diff
@@ -697,36 +722,28 @@ int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
...
-		casdsk_exp_obj_unlock(dsk);
+	if (dsk->openers > 0) {
+		mutex_unlock(&dsk->openers_lock);
 		return -EBUSY;
 	}
 
+	dsk->claimed = true;
+	mutex_unlock(&dsk->openers_lock);
+
 	return 0;
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c:735]: yup, good one.

File: modules/cas_disk/exp_obj.c (hunk 17 above)
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c:546]: agreed, reworked

**Review by jfckm — APPROVED** [review]

File: configure.d/1_alloc_disk.conf [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+		*queue = blk_mq_init_queue(tag_set);
+		if (IS_ERR_OR_NULL(*queue)) {
+			put_disk(*gd);
+			return -ENOMEM;
+		}
+		(*gd)->queue = *queue;
+
+		return 0;
+        }" ;;
```
[the remark this answers is not in the document]
Comment (mmichal10) [reply configure.d/1_alloc_disk.conf:54]: Additionally, I haven't found any usage of `blk_cleanup_queue()` as a part of the rollback for `blk_mq_alloc_disk()` so I guess it should be called only if `blk_mq_init_queue()` was called.

File: modules/cas_disk/exp_obj.c [hunk 23]
```diff
@@ -574,23 +596,26 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
 
 error_set_geometry:
+	_casdsk_exp_obj_clear_dev_t(dsk);
+error_exp_obj_set_dev_t:
 	blk_cleanup_queue(exp_obj->queue);
-error_init_queue:
+	put_disk(gd);
```
Comment (arutk) [inline modules/cas_disk/exp_obj.c:602]: please verify whether this call sequence is valid in case of kernel 5.15 (where gendisk and queue are both allocated using  blk_mq_alloc_disk

File: modules/cas_disk/exp_obj.c [hunk 24]
```diff
@@ -683,6 +714,7 @@ EXPORT_SYMBOL(casdsk_exp_obj_activated);
 int casdsk_exp_obj_lock(struct casdsk_disk *dsk)
 {
 	struct casdsk_exp_obj *exp_obj;
+	int result;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:717]: How about initializing `result` with `-EBUSY`? This way you could remove the first branch of `if`

File: modules/cas_disk/exp_obj.c [hunk 25]
```diff
@@ -339,8 +339,40 @@ static void _casdsk_exp_obj_clear_dev_t(struct casdsk_disk *dsk)
 	}
 }
 
+static int _casdsk_exp_obj_open(struct block_device *bdev, fmode_t mode)
+{
+	struct casdsk_disk *dsk = bdev->bd_disk->private_data;
+	int result;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:345]: How about initializing `result` with `-ENAVAIL`? This way you could get rid of the first branch of `if`

File: configure.d/1_alloc_disk.conf (hunk 22 above)
[the remark this answers is not in the document]
Comment (kmajzero) [reply configure.d/1_alloc_disk.conf:54]: added the generation of the cleanup function to the configure scripts

File: configure.d/1_bdget_disk.conf [hunk 26]
```diff
@@ -16,6 +16,9 @@ check() {
     elif compile_module $cur_name "bdgrab(NULL);" "linux/blkdev.h"
     then
         echo $cur_name 2 >> $config_file_path
+    elif compile_module $cur_name "blkdev_get_by_dev(0, 0, NULL);" "linux/blkdev.h"
```
[the remark this answers is not in the document]
Comment (kmajzero) [reply configure.d/1_bdget_disk.conf:19]: removed

File: modules/cas_disk/exp_obj.c (hunk 23 above)
Comment (kmajzero) [reply modules/cas_disk/exp_obj.c:602]: cleanup function added to the configure scripts

File: modules/cas_disk/exp_obj.c [hunk 27]
```diff
@@ -569,19 +596,22 @@ int casdsk_exp_obj_create(struct casdsk_disk *dsk, const char *dev_name,
...
+error_init_kobject:
 	module_put(owner);
 	dsk->exp_obj->owner = NULL;
 error_module_get:
-error_init_kobject:
 	kfree(exp_obj->dev_name);
 error_kstrdup:
-	casdsk_exp_obj_free(dsk);
+	__casdsk_exp_obj_release(exp_obj);
+	dsk->exp_obj = NULL;
```
Comment (mmichal10) [inline modules/cas_disk/exp_obj.c:614]: This change belongs to the previous commit

**Review by mmichal10 — COMMENTED** [review]
Please clean up commit history. Currently the changes which belong to the first commit are spread between both commits which leads to confusion during review. You can either reorganize the commits or melt everything into one commit

**Review by mmichal10 — APPROVED** [review]

**Review by arutk — APPROVED** [review]
